### PR TITLE
executor: truncate column values from index KV during admin check

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/23178 

Problem Summary:

```sql
create table t (a text collate utf8mb4_general_ci not null, b int(11) not null, 
  primary key (a(10), b) clustered, key idx(a(2)) ) 
  default charset=utf8mb4 collate=utf8mb4_bin;
insert into t values ("aaa", 2);
```

The key-value contents are:
```
// sk = sortKey, p = prefixed
row record:     sk(aaa), 2              -> aaa
index record:   sk(p(aa)), {sk(aaa), 2} -> restore data(aaa)
```

The error message says that the column value is `aa` from the row record, which is inconsistent with the one `aaa` from the index record:
```
mysql> admin check table t;
ERROR 1105 (HY000): col a, handle &kv.CommonHandle{encoded:[]uint8{0x1, 0x0, 0x41, 0x0, 0x41, 0x0, 0x41, 0x0, 0x0, 0xfd, 0x3, 0x80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x2}, colEndOffsets:[]uint16{0xa, 0x13}}, index:types.Datum{k:0x5, decimal:0x0, length:0x0, i:0, collation:"utf8mb4_general_ci", b:[]uint8{0x61, 0x61, 0x61}, x:interface {}(nil)} != record:types.Datum{k:0x5, decimal:0x0, length:0x0, i:0, collation:"utf8mb4_general_ci", b:[]uint8{0x61, 0x61}, x:interface {}(nil)}
```

This case is special because column `a` exists in both index values and common handle.

Before reading the row record, `TruncateIndexValues` is invoked. So we should do the same work for the datum read from index record.

### What is changed and how it works?

What's Changed: Truncate column values from index key-value during `admin check`.

### Related changes

NA

### Check List 

Tests 

- Unit test
- Integration test

Side effects

**Note:**
**This PR relax the check of `ADMIN CHECK`. It is possible that the data is inconsistent but `ADMIN CHECK` shows nothing.**

### Release note 

- No release note

### CODE REVIEW

File: executor/distsql.go [hunk 1]
```diff
@@ -1051,17 +1051,19 @@ func (w *tableWorker) compareData(ctx context.Context, task *lookupTableTask, ta
...
+			sctx := w.idxLookup.ctx.GetSessionVars().StmtCtx
 			for i, val := range vals {
 				col := w.idxTblCols[i]
 				tp := &col.FieldType
-				ret := chunk.Compare(idxRow, i, &val)
-				if ret != 0 {
-					return errors.Errorf("col %s, handle %#v, index:%#v != record:%#v", col.Name, handle, idxRow.GetDatum(i, tp), val)
+				idxVal := idxRow.GetDatum(i, tp)
+				tablecodec.TruncateIndexValue(&idxVal, w.idxLookup.index.Columns[i], col.ColumnInfo)
+				if res, err := idxVal.CompareDatum(sctx, &val); err != nil || res != 0 {
```
Comment (coocood) [inline executor/distsql.go:1060]: Can we remove the compare err in the format if it is nil?

File: executor/distsql.go [hunk 2]
```diff
@@ -1051,17 +1051,19 @@ func (w *tableWorker) compareData(ctx context.Context, task *lookupTableTask, ta
...
+					return errors.Errorf("col %s, handle %#v, index:%#v != record:%#v, compare err:%#v", col.Name,
```
Comment (coocood) [inline executor/distsql.go:1061]: So we don't need to change unrelated test cases.

**Review by lzmhhh123 — APPROVED** [review]

**Pull request merged**